File: built-in plugins dsp/mb-rotary.cpp
/**
 * MB Rotary Speaker
 * Category : effect
 * Type     : chorus
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Leslie speaker simulation
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_ROTARY_H
#define MB_ROTARY_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbRotary : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-rotary";
    static constexpr const char* PLUGIN_NAME    = "MB Rotary Speaker";
    static constexpr const char* PLUGIN_TYPE    = "chorus";
    static constexpr const char* PLUGIN_CATEGORY = "effect";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float speed = 0.5f;  // range [0, 1]
    float depth = 0.7f;  // range [0, 1]
    float mix = 0.6f;  // range [0, 1]
    };

    MbRotary() = default;
    ~MbRotary() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.speed = std::clamp(params.speed, 0f, 1f);
        params.depth = std::clamp(params.depth, 0f, 1f);
        params.mix = std::clamp(params.mix, 0f, 1f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Rotary Speaker
        return input;
    }
};

#endif // MB_ROTARY_H

File: built-in plugins dsp/mb-rotary.ts
import type { PluginDefinition } from '../server/services/pluginHostService';

const MbRotaryPlugin: PluginDefinition = { id: 'mb-rotary', slug: 'mb-rotary', name: 'MB Rotary Speaker', category: 'effect', type: 'chorus', version: '1.0.0', description: 'Leslie speaker simulation', author: 'Max Booster', grade: 'A', parameters: [{ id: 'speed', name: 'Speed', type: 'float', defaultValue: 0.5, minValue: 0, maxValue: 1, automatable: true }, { id: 'depth', name: 'Depth', type: 'float', defaultValue: 0.7, minValue: 0, maxValue: 1, automatable: true }, { id: 'mix', name: 'Mix', type: 'float', defaultValue: 0.6, minValue: 0, maxValue: 1, automatable: true }], defaultPreset: { speed: 0.5, depth: 0.7, mix: 0.6 } };

export default MbRotaryPlugin;
